main.tf

```
@@ -0,0 +1,124 @@
+resource "aws_vpc" "webapp" {
+  cidr_block           = var.cidr
+  enable_dns_support   = var.enable_dns_support
+  enable_dns_hostnames = var.enable_dns_hostnames
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "${local.name}-${local.environment}"
+    }
+  )
+}
+resource "aws_subnet" "public" {
+  for_each = toset(var.public_subnets)
+
+  cidr_block              = each.value
+  availability_zone       = element(data.aws_availability_zones.available.names, index(var.public_subnets, each.value))
+  vpc_id                  = aws_vpc.webapp.id
+  map_public_ip_on_launch = true
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "public-${each.value}"
+    }
+  )
+}
+
+resource "aws_internet_gateway" "public" {
+  vpc_id = aws_vpc.webapp.id
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "${local.name}-igw"
+    }
+  )
+}
+
+resource "aws_route_table" "public" {
+  vpc_id = aws_vpc.webapp.id
+
+  route {
+    cidr_block = "0.0.0.0/0"
+    gateway_id = aws_internet_gateway.public.id
+  }
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "public-${aws_internet_gateway.public.id}"
+    }
+  )
+}
+
+resource "aws_route_table_association" "public" {
+  for_each = aws_subnet.public
+
+  subnet_id      = each.value.id
+  route_table_id = aws_route_table.public.id
+}
+
+resource "aws_subnet" "private" {
+  for_each = toset(var.private_subnets)
+
+  cidr_block        = each.value
+  availability_zone = element(data.aws_availability_zones.available.names, index(var.private_subnets, each.value))
+  vpc_id            = aws_vpc.webapp.id
+
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "private-${each.value}"
+    }
+  )
+}
+
+resource "aws_eip" "private" {
+  for_each = aws_subnet.public
+  domain   = "vpc"
+  tags = merge(
+    local.base_tags,
+    {
+      Name = "${local.name}-${each.value.availability_zone}"
+    }
+  )
+}
+
+resource "aws_nat_gateway" "private" {
+  for_each = toset(var.public_subnets)
```

If you want to create one for every public subnet, rather than loop over the thing you used to create your public subnets, it's better practice to just loop over the public subnets directly.
```suggestion
  for_each = aws_subnet.public
```

---

I am using a lookup function where both route tables and subnets share the same "Availablity Zone" tag to create an association now. 

`resource "aws_route_table_association" "private" {
  for_each = aws_subnet.private

  subnet_id = each.value.id

  # Ensure both route tables and subnets share the same "Availablity Zone" tag 
  route_table_id = lookup(local.route_table_map, each.value.tags["AZ"], null)

  # This prevents errors if a route table isn't found for a subnet
  depends_on = [aws_route_table.private]
}`